crates/merkle-tree/src/tree.rs

```diff
@@ -521,107 +521,134 @@ impl<H: FeltHash, const HEIGHT: usize> MerkleTree<H, HEIGHT> {
         }
     }
 
-    /// Generates a merkle-proof for a given `key`.
+    /// Single-key version of [`MerkleTree::get_proofs`].
+    pub fn get_proof(
+        root: u64,
+        storage: &impl Storage,
+        key: &BitSlice<u8, Msb0>,
+    ) -> anyhow::Result<Option<Vec<TrieNode>>> {
+        Self::get_proofs(root, storage, &[key]).map(|mut proofs| proofs.pop().flatten())
+    }
+
+    /// Generates merkle-proofs for a given list of `keys`.
     ///
-    /// Returns vector of [`TrieNode`] which form a chain from the root to the
-    /// key, if it exists, or down to the node which proves that the key
-    /// does not exist.
+    /// For each key, returns a vector of [`TrieNode`]s which form a chain from
+    /// the root to the key, if it exists, or down to the node which proves
+    /// that the key does not exist. If a node in the path is missing from
+    /// storage, `None` will be returned for that path.
     ///
-    /// The nodes are returned in order, root first.
+    /// The nodes are added to the proof in order, root first.
     ///
     /// Verification is performed by confirming that:
     ///   1. the chain follows the path of `key`, and
     ///   2. the hashes are correct, and
     ///   3. the root hash matches the known root
-    pub fn get_proof(
+    ///
+    /// Uses a cache to avoid repeated lookups.
+    pub fn get_proofs(
         root: u64,
         storage: &impl Storage,
-        key: &BitSlice<u8, Msb0>,
-    ) -> anyhow::Result<Option<Vec<TrieNode>>> {
-        // Manually traverse towards the key.
-        let mut nodes = Vec::new();
-
-        let mut next = Some(root);
-        let mut height = 0;
-        while let Some(index) = next.take() {
-            let Some(node) = storage.get(index).context("Resolving node")? else {
-                return Ok(None);
-            };
+        keys: &[&BitSlice<u8, Msb0>],
+    ) -> anyhow::Result<Vec<Option<Vec<TrieNode>>>> {
+        let mut node_cache: HashMap<u64, StoredNode> = HashMap::new();
+        let mut proofs = vec![];
+
+        'key_loop: for key in keys {
+            // Manually traverse towards the key.
+            let mut nodes = Vec::new();
+
+            let mut next = Some(root);
+            let mut height = 0;
+            while let Some(index) = next.take() {
+                let node = match node_cache.get(&index) {
+                    Some(node) => node.clone(),
+                    None => {
+                        let Some(node) = storage.get(index).context("Resolving node")? else {
+                            proofs.push(None);
```

Comment:
BTW: how is this even possible? I mean: we're effectively traversing a tree and got a storage index for the node (either the root directly or from another node). Shouldn't this be an error here?

To be clear, this is _not_ new behavior I'm just unsure why we need it here.

Reply:
Yeah, exactly. Pruning should never remove nodes that are still accessible as part of a tree whose root we still have.

Reply:
Let's tackle that in a separate issue / PR then.